contract/src/main/kotlin/io/provenance/scope/loan/utility/FundingDataValidation.kt

```diff
@@ -9,14 +10,30 @@ internal val fundingValidation: ContractEnforcementContext.(Funding) -> Unit = {
     if (funding.isSet()) {
         funding.status.takeIf { it.isSet() }?.also { fundingStatus ->
             requireThat(
-                fundingStatus.status.isNotBlank()                   orError "Funding status must not be blank",
-                fundingStatus.effectiveTime.isValidForFundingTime() orError "Funding status must have valid effective time",
+                (fundingStatus.status in listOf("UNFUNDED", "FUNDED", "INITIATED", "COMPLETED", "CANCELLED"))
+                    orError "Funding status must be valid",
+                fundingStatus.effectiveTime.isValidFundingTime()
+                    orError "Funding status must have valid effective time",
             )
+            when (fundingStatus.status) {
+                in listOf("COMPLETED", "FUNDED") -> {
+                    requireThat(
+                        funding.started.isValidFundingTime()   orError "Completed funding's start time must be valid",
+                        funding.completed.isValidFundingTime() orError "Completed funding's end time must be valid",
+                    )
+                    if (funding.started.isValidFundingTime() && funding.completed.isValidFundingTime()) {
+                        requireThat(
+                            (funding.completed.toInstant() >= funding.started.toInstant()) orError "Funding end time must be after start time",
+                        )
+                    }
+                }
+                !in listOf("UNFUNDED", "UNKNOWN") -> {
+                    requireThat(
+                        funding.started.isValidFundingTime() orError "Funding start time must be valid",
+                    )
+                }
+            }
```

Comment:
@rgipson-figure mind validating these lists?

My questions:

1. Would completed time be set in the `FUNDED` status or only `COMPLETED`?
2. Should we check dates in the `CANCELED` status? (Think you mentioned completed would be set, but not sure if started would be, or if it's worth validating)

Not sure if you two already worked this out.

Reply:
1. Just added a comment above but I don't believe there is a FUNDED status. Basing this off of the p8etoMam converter

https://github.com/FigureTechnologies/provenance-corelib/blob/cd0ab7691864148d69cd27699b21d6ad11cb0af0/core-p8e-domain/src/main/kotlin/io/p8e/loan/P8eToMetadataAssetModelConverter.kt#L690

Reply:
2. Might be worth checking `completed` is set for CANCELLED and COMPLETED status. It least nice to have the date

Reply:
> I don't believe there is a FUNDED status

To be clear, do you mean "COMPLETED"? I see "FUNDED" in that code you linked.